File: ExampleObjc/ViewController.m
//
//  ViewController.m
//  ExampleObjc
//
//  Created by Dave Hardiman on 07/03/2016.
//  Copyright © 2016 Ordnance Survey. All rights reserved.
//

#import "ViewController.h"
@import Routing;
@import MapKit;

@interface ViewController ()<MKMapViewDelegate>
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (strong, nonatomic) OSRoutingService *routingService;
@property (strong, nonatomic) NSMutableArray *tappedPoints;
@end

@implementation ViewController

- (NSString *)apiKey {
    return [NSString stringWithContentsOfURL:[NSBundle.mainBundle URLForResource:@"APIKEY" withExtension:nil]
                                    encoding:NSUTF8StringEncoding
                                       error:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.routingService = [[OSRoutingService alloc] initWithApiKey:self.apiKey vehicleType:[OSRoutingService carVehicleType] crs:nil];
    self.mapView.delegate = self;
    self.tappedPoints = NSMutableArray.array;

    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(mapTapped:)];
    [self.mapView addGestureRecognizer:tap];
}

#pragma mark - Map Tapped
- (void)mapTapped:(UITapGestureRecognizer *)tap {
    if (tap.state != UIGestureRecognizerStateEnded) {
        return;
    }
    CGPoint touchPoint = [tap locationInView:self.mapView];
    CLLocationCoordinate2D touchCoordinate = [self.mapView convertPoint:touchPoint toCoordinateFromView:self.mapView];
    MKPointAnnotation *pointAnnotation = [[MKPointAnnotation alloc] init];
    pointAnnotation.coordinate = touchCoordinate;
    [self.mapView addAnnotation:pointAnnotation];
    [self.tappedPoints addObject:[NSValue os_valueWithCoordinate:touchCoordinate]];
    [self updateRoute];
}

- (void)updateRoute {
    [self.routingService routeBetweenCoordinates:self.tappedPoints completion:^(OSRoute *_Nullable route, NSError *_Nullable error) {
        if (error) {
            NSLog(@"%@", error);
            return;
        }
        [self displayRoute:route];
    }];
}

- (void)displayRoute:(OSRoute *)route {
    NSArray *coordinateValues = route.coordinateValues;
    NSInteger numberOfPoints = coordinateValues.count;
    CLLocationCoordinate2D coordinates[numberOfPoints];
    for (NSInteger idx = 0; idx < numberOfPoints; idx++) {
        coordinates[idx] = [coordinateValues[idx] os_coordinateValue];
    }
    MKPolyline *line = [MKPolyline polylineWithCoordinates:coordinates count:numberOfPoints];
    [self.mapView addOverlay:line];
}

- (MKOverlayRenderer *)mapView:(MKMapView *)mapView rendererForOverlay:(id<MKOverlay>)overlay {
    MKPolylineRenderer *renderer = [[MKPolylineRenderer alloc] initWithPolyline:overlay];
    renderer.strokeColor = [UIColor purpleColor];
    renderer.lineWidth = 3;
    return renderer;
}

#pragma mark - Clear
- (IBAction)clear:(id)sender {
    [self.mapView.annotations enumerateObjectsUsingBlock:^(id<MKAnnotation>  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [self.mapView removeAnnotation:obj];
    }];
    [self.mapView.overlays enumerateObjectsUsingBlock:^(id<MKOverlay>  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [self.mapView removeOverlay:obj];
    }];
    [self.tappedPoints removeAllObjects];
}


@end

File: ExampleSwift/ViewController.swift
//
//  ViewController.swift
//  ExampleSwift
//
//  Created by Dave Hardiman on 03/03/2016.
//  Copyright © 2016 Ordnance Survey. All rights reserved.
//

import UIKit
import MapKit
import Routing
import CoreLocation

class ViewController: UIViewController {

    var apiKey: String {
        return NSBundle.mainBundle().URLForResource("APIKEY", withExtension: nil).flatMap { url -> String? in
            do { return try String(contentsOfURL: url) } catch { return nil }
            } ?? ""
    }

    @IBOutlet weak var mapView: MKMapView!

    var routingService: RoutingService!

    lazy var tappedPoints = [CLLocationCoordinate2D]()

    override func viewDidLoad() {
        super.viewDidLoad()
        routingService = RoutingService(apiKey: apiKey, vehicleType: .Car)
        mapView.delegate = self

        let tap = UITapGestureRecognizer(target: self, action: "mapTapped:")
        mapView.addGestureRecognizer(tap)
    }

}

extension ViewController: MKMapViewDelegate {
    func mapView(mapView: MKMapView, rendererForOverlay overlay: MKOverlay) -> MKOverlayRenderer {
        if let overlay = overlay as? MKPolyline {
            let renderer = MKPolylineRenderer(polyline: overlay)
            renderer.lineWidth = 3
            renderer.strokeColor = UIColor.redColor()
            return renderer
        }
        return MKOverlayRenderer(overlay: overlay)
    }
}

extension ViewController {
    func mapTapped(tap: UITapGestureRecognizer) {
        guard tap.state == .Ended else {
            return
        }
        let touchPoint = tap.locationInView(mapView)
        let touchCoordinate = mapView.convertPoint(touchPoint, toCoordinateFromView: mapView)

        let pointAnnotation = MKPointAnnotation()
        pointAnnotation.coordinate = touchCoordinate
        mapView.addAnnotation(pointAnnotation)
        tappedPoints.append(touchCoordinate)
        updateRoute()
    }

    func updateRoute() {
        routingService.routeBetween(coordinates: tappedPoints) { result in
            switch result {
            case .Success(let route):
                self.displayRoute(route)
            case .Failure(let error):
                print("Routing failed \(error)")
            }
        }
    }

    func displayRoute(route: Route) {
        var coordinates = route.coordinates
        let line = MKPolyline(coordinates: &coordinates , count: coordinates.count)
        mapView.addOverlay(line)
    }
}

extension ViewController {
    @IBAction func clear() {
        mapView.overlays.forEach { overlay in
            mapView.removeOverlay(overlay)
        }
        mapView.annotations.forEach { annotation in
            mapView.removeAnnotation(annotation)
        }
        tappedPoints.removeAll()
    }

    @IBAction func sourceChanged(sender: UISegmentedControl) {
        clear()
        switch sender.selectedSegmentIndex {
        case 0:
            routingService = RoutingService(apiKey: apiKey, vehicleType: .Car)
        default:
            routingService = RoutingService(apiKey: apiKey, vehicleType: .Foot)
        }
    }

}
